flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/hybrid/HybridSourceReader.java

```diff
@@ -0,0 +1,201 @@
+                }
+                // more data will be available from the next reader
+                return InputStatus.MORE_AVAILABLE;
+            }
+        }
+        return status;
+    }
+
+    @Override
+    public List<HybridSourceSplit<SplitT>> snapshotState(long checkpointId) {
+        this.lastCheckpointId = checkpointId;
+        List<SplitT> state = currentReader.snapshotState(checkpointId);
+        return wrappedSplits(currentSourceIndex, state);
+    }
+
+    public static <SplitT extends SourceSplit> List<HybridSourceSplit<SplitT>> wrappedSplits(
+            int readerIndex, List<SplitT> state) {
+        List<HybridSourceSplit<SplitT>> wrappedSplits = new ArrayList<>(state.size());
+        for (SplitT split : state) {
+            wrappedSplits.add(new HybridSourceSplit<>(readerIndex, split));
+        }
+        return wrappedSplits;
+    }
+
+    public static <SplitT extends SourceSplit> List<SplitT> unwrappedSplits(
+            List<HybridSourceSplit<SplitT>> splits) {
+        List<SplitT> unwrappedSplits = new ArrayList<>(splits.size());
+        for (HybridSourceSplit<SplitT> split : splits) {
+            unwrappedSplits.add(split.getWrappedSplit());
+        }
+        return unwrappedSplits;
+    }
+
+    @Override
+    public CompletableFuture<Void> isAvailable() {
+        return currentReader.isAvailable();
+    }
+
+    @Override
+    public void addSplits(List<HybridSourceSplit<SplitT>> splits) {
+        LOG.info(
+                "### Adding splits subtask={} sourceIndex={} {} {}",
+                readerContext.getIndexOfSubtask(),
+                currentSourceIndex,
+                currentReader,
+                splits);
+        List<SplitT> realSplits = new ArrayList<>(splits.size());
+        for (HybridSourceSplit split : splits) {
+            Preconditions.checkState(
+                    split.sourceIndex() == currentSourceIndex,
+                    "Split %s while current source is %s",
+                    split,
+                    currentSourceIndex);
+            realSplits.add((SplitT) split.getWrappedSplit());
+        }
+        currentReader.addSplits(realSplits);
+    }
+
+    @Override
+    public void notifyNoMoreSplits() {
+        currentReader.notifyNoMoreSplits();
+        LOG.debug(
+                "No more splits for reader subtask={} sourceIndex={} {}",
+                readerContext.getIndexOfSubtask(),
+                currentSourceIndex,
+                currentReader);
+    }
+
+    @Override
+    public void handleSourceEvents(SourceEvent sourceEvent) {
+        if (sourceEvent instanceof SwitchSourceEvent) {
+            SwitchSourceEvent sse = (SwitchSourceEvent) sourceEvent;
+            LOG.debug(
+                    "Switch source event: subtask={} sourceIndex={}",
+                    readerContext.getIndexOfSubtask(),
+                    sse.sourceIndex());
+            setCurrentReader(sse.sourceIndex());
```

Comment:
IMO, the source reader switching should be executed before sending `SourceReaderFinishedEvent`, not receiving the `SwitchSourceEvent`.

Reply:
The enumerator only switches when all readers are done. Once the enumerator switches, it will send the `SwitchSourceEvent`

Reply:
I added to the javadoc to hopefully clarify the intent.